src/note.ts

```diff
@@ -595,13 +594,13 @@ export abstract class Note extends Tickable {
   }
 
   // Get all accidentals in the `ModifierContext`
-  getAccidentals(): Accidental[] {
-    return this.checkModifierContext().getMembers(Accidental.CATEGORY) as Accidental[];
+  getAccidentals(): Modifier[] {
```

Comment:
Thanks Rodrigo. This changes the semantics enough not to make these helpers useful. But that said, it seems like there's a bunch of options here:

- Remove it entirely, or,
- Replace this with a generic helper `getModifiers(type: string) -> Modifier[]`, or,
- longer term, start separating the interfaces and implementations.

Reply:
And what option do you prefer in the first instance?

Reply:
I prefer the second one, and moving `getAccidentals(note: Note): Accidental[]` to a static method in `accidentals.ts`.

Reply:
Done for `getAccidentals` and `getDots`.
@0xfe two questions related to the logic above:
- Do `addAccidental`, `addAnnotation` and `addArticulartion` make sense as helpers?
- Should `addDot` and `addDotToAll` be static funcstions of `Dot`?

Reply:
1) I think we should remove them from here, and move them into their respective classes.
2) Yes, agreed. (same as 1)

I think we can keep these APIs consistent (including the new one for `Parenthesis`.) I prefer the calls like:

```
Parenthesis.buildAndAttach(note)
Dot.buildAndAttach(note, index)
Accidental.buildAndAttach(note, index)
...
```

Reply:
there are two issues: 
- the parameters order of `addAccidental`, `addAnnotation` and `addArticulation` differ from the one in `addModifier` :( I would sugeest to leave untouched perhaps with a deprecated statement.
- while `Parenthesis.BuildAndAttach` and `Dot.buildAndAttach` are fine, `Accidenta.buildAndAttached` would require an additional parameter `type` used in the constructor. I would suggest just to do it for `Dot` and `Parenthesis`
- what about `addDotToAll`? I suggest `Dot.buildAndAttachToAll` this would make sense for Parenthesis as well.

Reply:
Looking into the dots details, I have realised that we will have to change code like (example from tests)
```
const notes = [
    staveNote({ keys: ['f/4'], duration: '4', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['f/4'], duration: '8', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['f/4'], duration: '16', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['f/4'], duration: '32', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['f/4'], duration: '64', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['f/4'], duration: '128', stem_direction: Stem.UP })
      .addDotToAll()
      .addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '4', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '8', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '16', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '32' }).addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '64', stem_direction: Stem.UP }).addDotToAll(),
    staveNote({ keys: ['g/4'], duration: '128', stem_direction: Stem.UP })
      .addDotToAll()
      .addDotToAll(),
  ];
```
into 
```
const notes = [
    staveNote({ keys: ['f/4'], duration: '4', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '8', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '16', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '32', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '64', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '128', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '4', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '8', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '16', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '32' }),
    staveNote({ keys: ['g/4'], duration: '64', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '128', stem_direction: Stem.UP }),
  ];
Dot.buildAndAttachToAll(notes[0]);
Dot.buildAndAttachToAll(notes[1]);
Dot.buildAndAttachToAll(notes[2]);
Dot.buildAndAttachToAll(notes[3]);
Dot.buildAndAttachToAll(notes[4]);
Dot.buildAndAttachToAll(notes[5]);
Dot.buildAndAttachToAll(notes[5]);
Dot.buildAndAttachToAll(notes[6]);
Dot.buildAndAttachToAll(notes[7]);
Dot.buildAndAttachToAll(notes[8]);
Dot.buildAndAttachToAll(notes[9]);
Dot.buildAndAttachToAll(notes[10]);
Dot.buildAndAttachToAll(notes[11]);
Dot.buildAndAttachToAll(notes[11]);
```
Do we want to do that? The code is more complex to understand.

Reply:
Yeah, I think it actually makes everything cleaner. There's no reason for `StaveNote` to know about `Dot`s, but `Dot`s must know about `StaveNotes`. The user does have to write more code, but the API becomes a bit more consistent, I think. I'd prefer the helpers to come from `Factory`, `System`, etc., and this core API to be sort of "low level."

Also, I think we could remove a lot of circular dependencies with this pattern.

Re: your comments:

1) The parameter orders were all jumbled, and we settled on `addModifier`, and wanted to remove the `addAccidental` etc. If you look at the `addModifier` code, you'll see the checks for inverse parameters (because even the `add*` methods were inversed.) I think it would be good to just get rid of those functions (or replace them with a `console.log` that says it's deprecated.)

2) I went through all the uses of accidental and dot as examples, and we can really remove so much there and clean up the dependencies. I think a more consistent API could be:

### elementInstance.attach(target, options)

```
let accidental = new Accidental(...);
accidental.attach(note, {index: 2});

let dot = new Dot();
dot.attach(note, {index: 2});
dot.attach(note, {all: true});
```

### elementClass.buildAndAttach(target, options)

```
Dot.buildAndAttach(note, {index: 0});
Dot.buildAndAttach(note, {all: true});
Parenthesis.buildAndAttach(note);
```

So `Note` can still have `addDot`, but it does not need a dependency on `Dot()`

### elementClass.getAttached(target)

```
// Replace note.getDots()
Dot.getAttached(note);

// Returns all accidentals 
Accidental.getAttached(note);
```

What do you think? If that makes sense, we can start with something like `Accidental` or `Dot` and see if we can take it out of the dependencies this way?

Reply:
I think that we will reduce the dependencies this way, no doubt, but I have to play with it a little bit to see the consequences. 
I do like the prototypes you suggested.

Reply:
q0xfe please move to #1270 while I look into that. @AaronDavidNewman needs to get that PR merged to continue resolving #411

Reply:
`addModifier` reason for the mess. `Note.addModifier` was changed from `(index, modifier)` to `(modifier,index)` during the TS migration due to the fact that `Tickable` also has a `addModifier` that was not signature compatible with `(index, modifier)`
@0xfe @ronyeh I would suggest to revert this change and rename the function in `Tickable` as the first step. What do you think?

Reply:
I have reverted the order of parameters. I think this will ease the migrations and there is no good reason for it. I will easily remove `addAnnotation`, `addAccidental` and `addArticulation`

Reply:
Removed

Reply:
I will go now for the buildAndAttach change. I believe that this one will reduce the number of dependencies. @0xfe Before I do it, would it make sense to change the signature to allow more than one note?
```
Dot.buildAndAttach([note], {index: 0});
Dot.buildAndAttach([note], {all: true});
Parenthesis.buildAndAttach([note]);
```
Normally parentheses and dots apply to more than one note. The example above would be:
```
const notes = [
    staveNote({ keys: ['f/4'], duration: '4', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '8', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '16', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '32', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '64', stem_direction: Stem.UP }),
    staveNote({ keys: ['f/4'], duration: '128', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '4', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '8', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '16', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '32' }),
    staveNote({ keys: ['g/4'], duration: '64', stem_direction: Stem.UP }),
    staveNote({ keys: ['g/4'], duration: '128', stem_direction: Stem.UP }),
  ];
Dot.buildAndAttach([notes], {all: true});
Dot.buildAndAttach([notes[5], notes[11]], {all: true});
``

Reply:
I have followed the proposal:
```
Dot.buildAndAttach([note], {index: 0});
Dot.buildAndAttach([note], {all: true});
Parenthesis.buildAndAttach([note]);
```
**and 6 additional circular dependencies are gone!**
**No visual differences!**

My thoughts after playing a bit:

> ### elementInstance.attach(target, options)
> ```
> let accidental = new Accidental(...);
> accidental.attach(note, {index: 2});
> 
> let dot = new Dot();
> dot.attach(note, {index: 2});
> dot.attach(note, {all: true});
> ```

we can do almost the same with `Note.addModifier` I do not see a big benefit here in regards to circular dependencies.

> ### elementClass.buildAndAttach(target, options)
> ```
> Dot.buildAndAttach(note, {index: 0});
> Dot.buildAndAttach(note, {all: true});
> Parenthesis.buildAndAttach(note);
> ```

done, very good against the circular dependencies.

> ### elementClass.getAttached(target)
> ```
> // Replace note.getDots()
> Dot.getAttached(note);
> 
> // Returns all accidentals 
> Accidental.getAttached(note);
> ```

we can do almost the same with `Note.getModifiersByType` I do not see a big benefit here in regards to circular dependencies.

Reply:
I would leave this PR as it is and make new ones if we want to resolve the remainig dependencies:
```
WARNING in Circular dependency detected:
src/accidental.ts -> src/gracenotegroup.ts -> src/beam.ts -> src/tuplet.ts -> src/formatter.ts -> src/modifiercontext.ts -> src/accidental.ts

WARNING in Circular dependency detected:
src/beam.ts -> src/tuplet.ts -> src/formatter.ts -> src/beam.ts

WARNING in Circular dependency detected:
src/factory.ts -> src/system.ts -> src/factory.ts

WARNING in Circular dependency detected:
src/formatter.ts -> src/beam.ts -> src/tuplet.ts -> src/formatter.ts

WARNING in Circular dependency detected:
src/glyph.ts -> src/tables.ts -> src/glyph.ts

WARNING in Circular dependency detected:
src/gracenotegroup.ts -> src/beam.ts -> src/tuplet.ts -> src/formatter.ts -> src/modifiercontext.ts -> src/accidental.ts -> src/gracenotegroup.ts

WARNING in Circular dependency detected:
src/modifiercontext.ts -> src/accidental.ts -> src/gracenotegroup.ts -> src/beam.ts -> src/tuplet.ts -> src/formatter.ts -> src/modifiercontext.ts

WARNING in Circular dependency detected:
src/notesubgroup.ts -> src/formatter.ts -> src/modifiercontext.ts -> src/notesubgroup.ts

WARNING in Circular dependency detected:
src/system.ts -> src/factory.ts -> src/system.ts

WARNING in Circular dependency detected:
src/tables.ts -> src/glyph.ts -> src/tables.ts

WARNING in Circular dependency detected:
src/tuplet.ts -> src/formatter.ts -> src/beam.ts -> src/tuplet.ts
```

Reply:
@0xfe ready to review